src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,560 @@
+                context.Trigger.Kind != CompletionTriggerKind.Insertion)
+            {
+                return;
+            }
+
+            var position = context.Position;
+            var (tree, stringToken) = await _language.TryGetTreeAndTokenAtPositionAsync(
+                context.Document, position, context.CancellationToken).ConfigureAwait(false);
+
+            var embeddedContext = new EmbeddedCompletionContext(this, context, tree, stringToken);
+            ProvideCompletions(embeddedContext);
+
+            if (embeddedContext.Items.Count == 0)
+            {
+                return;
+            }
+
+            foreach (var embeddedItem in embeddedContext.Items)
+            {
+                var change = embeddedItem.Change;
+                var textChange = change.TextChange;
+
+                var properties = ImmutableDictionary.CreateBuilder<string, string>();
+                properties.Add(StartKey, textChange.Span.Start.ToString());
+                properties.Add(LengthKey, textChange.Span.Length.ToString());
+                properties.Add(NewTextKey, textChange.NewText);
+                properties.Add(DescriptionKey, embeddedItem.FullDescription);
+                properties.Add(AbstractEmbeddedLanguageCompletionProvider.EmbeddedProviderName, this.Name);
+
+                if (change.NewPosition != null)
+                {
+                    properties.Add(NewPositionKey, change.NewPosition.ToString());
+                }
+
+                // Keep everything sorted in the order we just produced the items in.
+                var sortText = context.Items.Count.ToString("0000");
+                context.AddItem(CompletionItem.Create(
+                    displayText: embeddedItem.DisplayText,
+                    sortText: sortText,
+                    properties: properties.ToImmutable(),
+                    rules: s_rules,
+                    inlineDescription: "    " + embeddedItem.InlineDescription));
+            }
+
+            context.IsExclusive = true;
+        }
+
+        private void ProvideCompletions(EmbeddedCompletionContext context)
+        {
+            var position = context.Position;
+
+            var tree = context.Tree;
+            var stringToken = context.StringToken;
+            if (tree == null ||
+                position <= stringToken.SpanStart ||
+                position >= stringToken.Span.End)
+            {
+                return;
+            }
+
+            // First, act as if the user just inserted the previous character.  This will cause us
+            // to complete down to the set of relevant items based on that character. If we get
+            // anything, we're done and can just show the user those items.  If we have no items to
+            // add *and* the user was explicitly invoking completion, then just add the entire set
+            // of suggestions to help the user out.
```

Comment:
Isn't this just kinda reinventing filtering? What happens if this filtering happens based on the insertion, and then I backspace?